reactive/kubernetes_master.py

```diff
@@ -1720,19 +1721,35 @@ def configure_scheduler():
 
 def setup_basic_auth(password=None, username='admin', uid='admin',
                      groups=None):
-    '''Create the htacces file and the tokens.'''
+    '''Add or update an entry in /root/cdk/basic_auth.csv.'''
     root_cdk = '/root/cdk'
     if not os.path.isdir(root_cdk):
         os.makedirs(root_cdk)
     htaccess = os.path.join(root_cdk, 'basic_auth.csv')
     if not password:
         password = token_generator()
-    with open(htaccess, 'w') as stream:
-        if groups:
-            stream.write('{0},{1},{2},"{3}"'.format(password,
-                                                    username, uid, groups))
-        else:
-            stream.write('{0},{1},{2}'.format(password, username, uid))
+
+    newline = '{0},{1},{2}'.format(password, username, uid)
+    if groups:
+        newline += ',"{0}"'.format(groups)
+
+    # If we have an existing line for this username, update it, leaving
+    # lines for other users unchanged.
+    with fileinput.input(htaccess, inplace=True) as f:
+        user_written = False
+
+        for line in f:
+            line_user, line_uid = line.split(',')[1:3]
+            if (line_user, line_uid) == (username, uid):
+                print(newline)
+                user_written = True
+            else:
+                print(line.strip())
+
+    # If we didn't find an existing line for this username, add one.
+    if not user_written:
+        with open(htaccess, 'a') as f:
+            f.write(newline)
```

Comment:
Seems like it would have been a bit cleaner to nest this under the context manager and just use `print()`.

Reply:
Ah, yeah, that's what I tried first, but it doesn't work. You can't append to the file. I found it weird and unexpected, but apparently it's by design, since the docs say: "With mode you can specify which file mode will be passed to open(). It must be one of 'r', 'rU', 'U' and 'rb'."

Reply:
Actually, that doesn't really explain it. Clearly the file is being written. But regardless, I stepped through it in the debugger, and although the print call was made, the extra line was not written to the file. It seemed that with fileinput, you can write an existing line, or change an existing line, but you can't add a new line.

Reply:
Yeah, so I read the source for fileinput. When the loop through the lines finishes, the output file is closed, so you can't write any additional lines.

Reply:
Very strange.  It seems that fileinput with `inplace=True` only does the `stdout` redirect when actually iterating over the lines, rather than during the entire context manager block:

```python
In [22]: filepath.write_text('one\ntwo\nthree')                                                                                                                                                                                  
Out[22]: 13

In [23]: print(filepath.read_text())                                                                                                                                                                                             
one
two
three

In [24]: with fileinput.input(str(filepath), inplace=True) as f: 
    ...:     print('new - start') 
    ...:     for l in f: 
    ...:         print(l.strip()) 
    ...:         print(l.strip()) 
    ...:     print('new - end') 
    ...:                                                                                                                                                                                                                         
new - start
new - end

In [25]: print(filepath.read_text())                                                                                                                                                                                             
one
one
two
two
three
three


In [26]: 
```

I guess that makes sense from the perspective of it being able to process multiple files with a single loop.